File: src/caco/tui/widgets/wad_table.py
"""WAD list table widget."""

import asyncio

from rich.text import Text
from textual.binding import Binding
from textual.widgets import DataTable
from textual.message import Message

from caco import db
from caco.player import format_duration
from caco.tui.theme import get_status_color


class WadTable(DataTable):
    """DataTable for displaying WAD list with vim-style navigation."""

    BINDINGS = [
        Binding("j", "cursor_down", "Down", show=False),
        Binding("k", "cursor_up", "Up", show=False),
        Binding("g", "handle_g", "Top (gg)", show=False),
        Binding("G", "go_bottom", "Bottom", show=False),
        Binding("ctrl+d", "page_down", "Page Down", show=False),
        Binding("ctrl+u", "page_up", "Page Up", show=False),
        Binding("enter", "select_cursor", "Play", show=True),
    ]

    class WadSelected(Message):
        """Fired when cursor moves to a new WAD."""

        def __init__(self, wad_id: int | None) -> None:
            super().__init__()
            self.wad_id = wad_id

    class WadActivated(Message):
        """Fired when Enter is pressed on a WAD (to play it)."""

        def __init__(self, wad_id: int) -> None:
            super().__init__()
            self.wad_id = wad_id

    def __init__(self, **kwargs) -> None:
        super().__init__(**kwargs)
        self._wads: list[dict] = []
        self._wad_id_to_row: dict[int, int] = {}
        self._g_pressed = False
        self._g_timeout_task: asyncio.Task | None = None
        # Unified stats map: {wad_id: {playtime, last_played, session_count, times_beaten}}
        self._stats_map: dict[int, dict] = {}

    def on_mount(self) -> None:
        """Set up the table columns."""
        self.cursor_type = "row"
        self.zebra_stripes = True

        # Add columns
        self.add_column("ID", key="id", width=5)
        self.add_column("Title", key="title", width=30)
        self.add_column("Author", key="author", width=20)
        self.add_column("Status", key="status", width=12)
        self.add_column("Playtime", key="playtime", width=10)

    def load_wads(
        self,
        query: str | None = None,
        sort_by: str = "id",
        sort_desc: bool = False,
        include_deleted: bool = False,
    ) -> int:
        """Load WADs from database and populate table.

        Returns the number of WADs loaded.
        """
        self._wads = db.search_wads(
            query, sort_by=sort_by, sort_desc=sort_desc,
            include_deleted=include_deleted,
        )
        self._wad_id_to_row.clear()
        self.clear()

        if not self._wads:
            self._stats_map = {}
            return 0

        # Batch-fetch all stats in 2 queries on 1 connection
        wad_ids = [wad["id"] for wad in self._wads]
        self._stats_map = db.get_wad_stats_batch(wad_ids)

        for i, wad in enumerate(self._wads):
            wad_id = wad["id"]
            self._wad_id_to_row[wad_id] = i

            # Format playtime from unified stats map
            playtime = self._stats_map.get(wad_id, {}).get("playtime", 0)
            playtime_str = format_duration(playtime) if playtime else "-"

            # Status with color styling using Rich Text
            status = wad["status"]
            color = get_status_color(status)
            status_text = Text(status, style=color) if color else status

            self.add_row(
                str(wad_id),
                wad["title"],
                wad["author"] or "-",
                status_text,
                playtime_str,
                key=str(wad_id),
            )

        # Notify about initial selection
        if self._wads:
            self.post_message(self.WadSelected(self._wads[0]["id"]))

        return len(self._wads)

    def get_wad_stats(self, wad_id: int) -> dict:
        """Get pre-fetched stats for a WAD (avoids extra DB queries).

        Returns dict with: playtime, last_played, times_beaten, session_count.
        """
        defaults = {"playtime": 0, "last_played": None, "times_beaten": 0, "session_count": 0}
        return self._stats_map.get(wad_id, defaults)

    def update_row(self, wad_id: int) -> bool:
        """Update a single row in-place without full table reload.

        Returns True if the row was updated, False if wad_id not found in table.
        """
        if wad_id not in self._wad_id_to_row:
            return False

        row_idx = self._wad_id_to_row[wad_id]
        wad = db.get_wad(wad_id)
        if not wad:
            return False

        # Update the cached wad data
        self._wads[row_idx] = wad

        # Refresh stats for this single WAD (1 call, 2 queries)
        self._stats_map.update(db.get_wad_stats_batch([wad_id]))

        # Update cells in the DataTable
        row_key = str(wad_id)
        playtime = self._stats_map.get(wad_id, {}).get("playtime", 0)
        playtime_str = format_duration(playtime) if playtime else "-"

        status = wad["status"]
        color = get_status_color(status)
        status_text = Text(status, style=color) if color else status

        self.update_cell(row_key, "title", wad["title"])
        self.update_cell(row_key, "author", wad["author"] or "-")
        self.update_cell(row_key, "status", status_text)
        self.update_cell(row_key, "playtime", playtime_str)

        return True

    def get_selected_wad_id(self) -> int | None:
        """Get the currently selected WAD ID."""
        if self.cursor_row is not None and 0 <= self.cursor_row < len(self._wads):
            return self._wads[self.cursor_row]["id"]
        return None

    def on_data_table_row_highlighted(self, event: DataTable.RowHighlighted) -> None:
        """Handle row highlight change."""
        if event.cursor_row is not None and 0 <= event.cursor_row < len(self._wads):
            wad_id = self._wads[event.cursor_row]["id"]
            self.post_message(self.WadSelected(wad_id))
        else:
            self.post_message(self.WadSelected(None))

    def on_data_table_row_selected(self, event: DataTable.RowSelected) -> None:
        """Handle Enter key on a row."""
        wad_id = self.get_selected_wad_id()
        if wad_id is not None:
            self.post_message(self.WadActivated(wad_id))

    async def action_cursor_down(self) -> None:
        """Move cursor down (j key)."""
        self._g_pressed = False
        if self.cursor_row is not None and self.cursor_row < len(self._wads) - 1:
            self.move_cursor(row=self.cursor_row + 1)

    async def action_cursor_up(self) -> None:
        """Move cursor up (k key)."""
        self._g_pressed = False
        if self.cursor_row is not None and self.cursor_row > 0:
            self.move_cursor(row=self.cursor_row - 1)

    async def action_page_down(self) -> None:
        """Move cursor down by half page (Ctrl+d)."""
        self._g_pressed = False
        if self.cursor_row is not None:
            # Move by ~10 rows or to end
            new_row = min(self.cursor_row + 10, len(self._wads) - 1)
            self.move_cursor(row=new_row)

    async def action_page_up(self) -> None:
        """Move cursor up by half page (Ctrl+u)."""
        self._g_pressed = False
        if self.cursor_row is not None:
            # Move up by ~10 rows or to start
            new_row = max(self.cursor_row - 10, 0)
            self.move_cursor(row=new_row)

    async def action_go_top(self) -> None:
        """Go to top of list (gg)."""
        self._g_pressed = False
        if self._wads:
            self.move_cursor(row=0)

    async def action_go_bottom(self) -> None:
        """Go to bottom of list (G)."""
        self._g_pressed = False
        if self._wads:
            self.move_cursor(row=len(self._wads) - 1)

    def action_handle_g(self) -> None:
        """Handle 'g' key press for gg motion."""
        if self._g_pressed:
            # Second g - go to top
            self._g_pressed = False
            if self._g_timeout_task:
                self._g_timeout_task.cancel()
            self.run_worker(self.action_go_top())
        else:
            # First g - wait for second
            self._g_pressed = True
            self._g_timeout_task = asyncio.create_task(self._g_timeout())

    async def _g_timeout(self) -> None:
        """Reset g state after timeout."""
        await asyncio.sleep(0.5)
        self._g_pressed = False

    def reset_g_state(self) -> None:
        """Reset g key state."""
        self._g_pressed = False

File: crates/caco-tui/src/widgets/wad_table.rs
use std::collections::HashMap;
use std::time::Instant;

use caco_core::db::models::WadRecord;
use caco_core::db::query::search_wads;
use caco_core::db::sessions::{WadStats, get_wad_stats_batch};
use caco_core::player::format_duration;
use ratatui::layout::{Constraint, Rect};
use ratatui::style::{Modifier, Style};
use ratatui::widgets::{Cell, Row, Table, TableState};
use ratatui::Frame;
use rusqlite::Connection;

use crate::theme;

/// State for the WAD table widget.
pub struct WadTableState {
    pub wads: Vec<WadRecord>,
    pub stats_map: HashMap<i64, WadStats>,
    pub table_state: TableState,
    id_to_index: HashMap<i64, usize>,
    /// For gg (double-g to go to top): tracks when first g was pressed.
    pub g_pressed: Option<Instant>,
}

impl Default for WadTableState {
    fn default() -> Self {
        Self::new()
    }
}

impl WadTableState {
    pub fn new() -> Self {
        Self {
            wads: Vec::new(),
            stats_map: HashMap::new(),
            table_state: TableState::default(),
            id_to_index: HashMap::new(),
            g_pressed: None,
        }
    }

    /// Load WADs from the database with the given query and sort parameters.
    pub fn load_wads(
        &mut self,
        conn: &Connection,
        query: Option<&str>,
        sort_by: Option<&str>,
        sort_desc: bool,
        include_deleted: bool,
    ) -> usize {
        match search_wads(conn, query, sort_by, sort_desc, include_deleted, 0) {
            Ok(wads) => {
                let ids: Vec<i64> = wads.iter().map(|w| w.id).collect();
                self.stats_map = get_wad_stats_batch(conn, &ids).unwrap_or_default();
                self.id_to_index = ids
                    .iter()
                    .enumerate()
                    .map(|(i, &id)| (id, i))
                    .collect();
                let count = wads.len();
                self.wads = wads;

                // Preserve selection or select first
                if count > 0 {
                    let current = self.table_state.selected().unwrap_or(0);
                    if current >= count {
                        self.table_state.select(Some(count - 1));
                    } else {
                        self.table_state.select(Some(current));
                    }
                } else {
                    self.table_state.select(None);
                }

                count
            }
            Err(_) => {
                self.wads.clear();
                self.stats_map.clear();
                self.id_to_index.clear();
                self.table_state.select(None);
                0
            }
        }
    }

    /// Refresh a single WAD's data in-place.
    pub fn update_row(&mut self, conn: &Connection, wad_id: i64) -> bool {
        if let Some(&idx) = self.id_to_index.get(&wad_id) {
            if let Ok(Some(mut wad)) = caco_core::db::wads::get_wad(conn, wad_id, true) {
                let _ = caco_core::db::connection::attach_tags(conn, &mut wad);
                self.wads[idx] = wad;
                if let Ok(stats) = get_wad_stats_batch(conn, &[wad_id]) {
                    if let Some(s) = stats.into_iter().next() {
                        self.stats_map.insert(s.0, s.1);
                    }
                }
                return true;
            }
        }
        false
    }

    /// Get the currently selected WAD ID.
    pub fn selected_wad_id(&self) -> Option<i64> {
        self.table_state
            .selected()
            .and_then(|i| self.wads.get(i))
            .map(|w| w.id)
    }

    /// Get the currently selected WAD record.
    pub fn selected_wad(&self) -> Option<&WadRecord> {
        self.table_state
            .selected()
            .and_then(|i| self.wads.get(i))
    }

    /// Get stats for the currently selected WAD.
    pub fn selected_stats(&self) -> Option<&WadStats> {
        self.selected_wad_id()
            .and_then(|id| self.stats_map.get(&id))
    }

    /// Move selection down by one.
    pub fn next(&mut self) {
        if self.wads.is_empty() {
            return;
        }
        let i = match self.table_state.selected() {
            Some(i) => (i + 1).min(self.wads.len() - 1),
            None => 0,
        };
        self.table_state.select(Some(i));
    }

    /// Move selection up by one.
    pub fn previous(&mut self) {
        if self.wads.is_empty() {
            return;
        }
        let i = match self.table_state.selected() {
            Some(i) => i.saturating_sub(1),
            None => 0,
        };
        self.table_state.select(Some(i));
    }

    /// Move to first row.
    pub fn first(&mut self) {
        if !self.wads.is_empty() {
            self.table_state.select(Some(0));
        }
    }

    /// Move to last row.
    pub fn last(&mut self) {
        if !self.wads.is_empty() {
            self.table_state.select(Some(self.wads.len() - 1));
        }
    }

    /// Page down (half screen).
    pub fn page_down(&mut self, visible_rows: usize) {
        if self.wads.is_empty() {
            return;
        }
        let half = visible_rows / 2;
        let i = match self.table_state.selected() {
            Some(i) => (i + half).min(self.wads.len() - 1),
            None => 0,
        };
        self.table_state.select(Some(i));
    }

    /// Page up (half screen).
    pub fn page_up(&mut self, visible_rows: usize) {
        if self.wads.is_empty() {
            return;
        }
        let half = visible_rows / 2;
        let i = match self.table_state.selected() {
            Some(i) => i.saturating_sub(half),
            None => 0,
        };
        self.table_state.select(Some(i));
    }

    /// Get WAD record by ID from the cached list.
    pub fn get_wad_by_id(&self, wad_id: i64) -> Option<&WadRecord> {
        self.id_to_index
            .get(&wad_id)
            .and_then(|&i| self.wads.get(i))
    }

    /// Get stats for a WAD by ID.
    pub fn get_wad_stats(&self, wad_id: i64) -> Option<&WadStats> {
        self.stats_map.get(&wad_id)
    }
}

/// Render the WAD table.
pub fn render_wad_table(state: &mut WadTableState, frame: &mut Frame, area: Rect) {
    let header = Row::new(vec![
        Cell::from("ID").style(Style::default().add_modifier(Modifier::BOLD)),
        Cell::from("Title").style(Style::default().add_modifier(Modifier::BOLD)),
        Cell::from("Author").style(Style::default().add_modifier(Modifier::BOLD)),
        Cell::from("Status").style(Style::default().add_modifier(Modifier::BOLD)),
        Cell::from("Playtime").style(Style::default().add_modifier(Modifier::BOLD)),
    ])
    .height(1);

    let rows: Vec<Row> = state
        .wads
        .iter()
        .map(|wad| {
            let stats = state.stats_map.get(&wad.id);
            let playtime = stats
                .map(|s| {
                    if s.playtime > 0 {
                        format_duration(s.playtime)
                    } else {
                        String::new()
                    }
                })
                .unwrap_or_default();

            let status_str = theme::status_display(&wad.status);
            let status_style = theme::status_style(&wad.status);

            let deleted_style = if wad.deleted_at.is_some() {
                Style::default().add_modifier(Modifier::DIM)
            } else {
                Style::default()
            };

            Row::new(vec![
                Cell::from(wad.id.to_string()).style(deleted_style),
                Cell::from(wad.title.clone()).style(deleted_style),
                Cell::from(wad.author.clone().unwrap_or_default()).style(deleted_style),
                Cell::from(status_str).style(status_style),
                Cell::from(playtime).style(deleted_style),
            ])
        })
        .collect();

    let widths = [
        Constraint::Length(5),
        Constraint::Min(20),
        Constraint::Length(20),
        Constraint::Length(16),
        Constraint::Length(10),
    ];

    let table = Table::new(rows, widths)
        .header(header)
        .row_highlight_style(theme::highlight_style());

    frame.render_stateful_widget(table, area, &mut state.table_state);
}
